nni/nas/evaluator/evaluator.py

```diff
@@ -3,58 +3,389 @@
+            return mutated_obj
+
+        return obj
+
+    @staticmethod
+    def expand_trace_kwargs(obj: Any) -> Iterator[Mutable]:
+        if not is_traceable(obj, must_be_instance=True):
+            raise TypeError(f'{obj} is not traceable.')
+        for pos, param in enumerate(obj.trace_args):
+            if isinstance(param, Mutable):
+                raise ValueError(f'We currently do not support mutable parameters in positional arguments: {obj} (position: {pos})')
+        for param in obj.trace_kwargs.values():
+            if isinstance(param, Mutable):
+                yield param
+            # Recursively yield all the nested parameters
+            if is_traceable(param, must_be_instance=True):
+                yield from MutableEvaluator.expand_trace_kwargs(param)
+
+    def freeze(self, sample: Sample) -> FrozenEvaluator | MutableEvaluator:
+        """Upon freeze, :class:`MutableEvaluator` will freeze all the mutable parameters
+        (as well as nested parameters),
+        and return a :class:`FrozenEvaluator`.
+
+        The evaluator will not be fully initialized to save the memory,
+        especially when parameters contain large objects such as datasets.
+        To use the evaluator, call :meth:`FrozenEvaluator.get` to get the full usable evaluator.
+
+        Returns
+        -------
+        The frozen evaluator.
+        """
+        frozen_evaluator = self.freeze_traceable_object(self, sample)
+        assert isinstance(frozen_evaluator, (MutableEvaluator, SerializableObject)), \
+            'The evaluator must still be a MutableEvaluator or SerializableObject after freezing.'
+        if type(frozen_evaluator) is SerializableObject:
+            return FrozenEvaluator(frozen_evaluator)
+        else:
+            assert isinstance(frozen_evaluator, MutableEvaluator)
+            return frozen_evaluator
+
+    def leaf_mutables(self, is_leaf: Callable[[Mutable], bool]) -> Iterable[LabeledMutable]:
+        for param in self.expand_trace_kwargs(self):
+            yield from param.leaf_mutables(is_leaf)
+
+    def check_contains(self, sample: Sample) -> SampleValidationError | None:
+        for mutable in self.expand_trace_kwargs(self):
+            exception = mutable.check_contains(sample)
+            if exception is not None:
+                exception.paths.append(str(mutable))
+                return exception
+        return None
+
+    def is_mutable(self) -> bool:
+        """Whether some arguments of the evaluator are mutable.
+
+        Although the evaluator is mutable, it may contain no mutable parameters,
+        i.e., all its parameters (including nested ones) are fixed values.
+        Return false if there is none.
         """
-        Subclass implements ``_dump`` for their own serialization.
-        They should return a dict, with a key ``type`` which equals ``self.__class__``,
-        and optionally other keys.
+        for _ in self.expand_trace_kwargs(self):
+            return True
```

Comment:
🤣

Reply:
It's a clever way to check whether an iterator is not empty, without expanding it.